pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,468 @@
+		glog.Infof("LB already exists and doesn't need update for service %s", namespacedName)
+		return nil, notRetryable
+	}
+	if cachedService != nil {
+		// If the service already exists but needs to be updated, delete it so that
+		// we can recreate it cleanly.
+		if cachedService.Spec.CreateExternalLoadBalancer {
+			glog.Infof("Deleting existing load balancer for service %s that needs an updated load balancer.", namespacedName)
+			if err := s.ensureLBDeleted(cachedService); err != nil {
+				return err, retryable
+			}
+		}
+	} else {
+		// If we don't have any cached memory of the load balancer and it already
+		// exists, optimistically consider our work done.
+		// TODO: If we could read the spec of the existing load balancer, we could
+		// determine if an update is necessary.
+		exists, err := s.balancer.TCPLoadBalancerExists(s.loadBalancerName(service), s.zone.Region)
+		if err != nil {
+			return fmt.Errorf("Error getting LB for service %s", namespacedName), retryable
+		}
+		if exists && len(service.Spec.PublicIPs) == 0 {
+			// The load balancer exists, but we apparently don't know about its public
+			// IPs, so just delete it and recreate it to get back to a sane state.
+			// TODO: Ideally the cloud provider interface would return the IP for us.
+			glog.Infof("Deleting old LB for service with no public IPs %s", namespacedName)
+			if err := s.ensureLBDeleted(service); err != nil {
+				return err, retryable
+			}
+		} else if exists {
+			// TODO: Better handle updates for non-cached services, this is optimistic.
+			glog.Infof("LB already exists for service %s", namespacedName)
+			return nil, notRetryable
+		}
+	}
+
+	if !service.Spec.CreateExternalLoadBalancer {
+		glog.Infof("Not creating LB for service %s that doesn't want one.", namespacedName)
+		return nil, notRetryable
+	}
+
+	glog.V(2).Infof("Creating LB for service %s", namespacedName)
+
+	// The load balancer doesn't exist yet, so create it.
+	publicIPstring := fmt.Sprint(service.Spec.PublicIPs)
+	err := s.createExternalLoadBalancer(service)
+	if err != nil {
+		return fmt.Errorf("failed to create external load balancer for service %s: %v", namespacedName, err), retryable
+	}
+
+	if publicIPstring == fmt.Sprint(service.Spec.PublicIPs) {
+		glog.Infof("Not persisting unchanged service to registry.")
+		return nil, notRetryable
+	}
+
+	// If creating the load balancer succeeded, persist the updated service.
+	if err = s.persistUpdate(service); err != nil {
+		return fmt.Errorf("Failed to persist updated publicIPs to apiserver, even after retries. Giving up: %v", err), notRetryable
+	}
+	return nil, notRetryable
+}
+
+func (s *ServiceController) persistUpdate(service *api.Service) error {
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		_, err = s.kubeClient.Services(service.Namespace).Update(service)
+		if err == nil {
+			return nil
+		}
+		if errors.IsNotFound(err) {
+			glog.Infof("Not persisting update to service that no longer exists: %v", err)
```

Comment:
What's the error? Unless If the service no longer exists, it must have been deleted. If it was deleted, we don't want to re-create it, we just want to bail out so that we can process the deletion event. I've tried explaining this in a new comment, but please tell me if I'm missing something or this isn't clear.